src/main/ruby/truffleruby/core/posix.rb

```diff
@@ -425,11 +425,62 @@ def self.read_string_native(io, length)
     end
   end
 
-  def self.read_to_buffer_native(io, length)
+  def self.execute_posix_read(io, buffer, length)
+    fd = io.fileno
+    return Truffle::POSIX.read(fd, buffer, length) unless io.timeout
+
+    deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + io.timeout
+
+    loop do
+      current_timeout = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      raise IO::TimeoutError if current_timeout < 0
+
+      poll_result = Truffle::IOOperations.poll(io, Truffle::IOOperations::POLLIN, current_timeout)
+      if poll_result == 0
+        raise IO::TimeoutError
+      elsif poll_result == -1
+        Errno.handle_errno(Errno.errno)
+      end
+
+      if (bytes_read = Truffle::POSIX.read(fd, buffer, length)) == -1
+        continue if Errno.errno == Errno::EAGAIN
```

Comment:
There is no `continue` in Ruby, but there is `next`

Reply:
Ah, a c moment.